deps/rabbit/src/tcp_listener.erl

```diff
@@ -79,9 +87,37 @@ handle_cast(_Msg, State) ->
 handle_info(_Info, State) ->
     {noreply, State}.
 
-terminate(_Reason, #state{on_shutdown = {M,F,A}, label=Label, ip=IPAddress, port=Port}) ->
+terminate(_Reason, #state{label = Label, ip = IPAddress, port = Port} = State0) ->
+    #state{on_shutdown = {M, F, A}} = deobfuscate_state(State0),
     logger:info("stopped ~s on ~s:~p", [Label, rabbit_misc:ntoab(IPAddress), Port]),
-    apply(M, F, A ++ [IPAddress, Port]).
+    try
+        apply(M, F, A ++ [IPAddress, Port])
+    catch _:Error ->
+        logger:error("Failed to stop listener ~s on ~s:~p: ~p",
+                     [Label, rabbit_misc:ntoab(IPAddress), Port, Error])
+    end.
 
 code_change(_OldVsn, State, _Extra) ->
     {ok, State}.
+
+obfuscate_state(#state{on_startup = OnStartup, on_shutdown = OnShutdown} = State) ->
+    State#state{
+        on_startup  = credentials_obfuscation:encrypt(term_to_binary(OnStartup)),
+        on_shutdown = credentials_obfuscation:encrypt(term_to_binary(OnShutdown))
```

Comment:
Instead of doing encrypt and TTB you can just do `fun(IPAddress, Port) -> apply(M, F, A ++ [IPAddress, Port]) end` as the MFAs will be captured and hidden inside the fun.

Reply:
This would be done in the init function, and the apply in terminate can be replaced with `OnShutdownFun(IPAddress, Port)`.